Voltage divider

ualbertagreen opened the pull request:
This PR adds initialization macro and tests to the voltage-divider binding.

ualbertagreen committed:
drivers: adc: voltage divider

adds DT macro initializer and scaling function for voltage divider.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen committed:
test: adc: voltage divider

adds tests for the voltage divider binding macro and rescaling function.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen commented:
@galak @fabiobaltieri Hi, I've created a PR for macro and tests for the voltage-divider binding in this PR. Let me know what you guys think and if it looks good I'll stack the current-sensor on top of this PR.

fabiobaltieri commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,73 @@
+static inline int divider_scale_voltage_dt(const struct voltage_divider_dt_spec *spec,
+					   int32_t *v_to_v)
+{
+	/* use as resistance based voltage sensor if "full_ohms" is 0 */
```
I think this refers to sensors like PTC/NTC resistors, if that's the case maybe we should return an error if full_ohms is not set, and have a second translation function that instead of the full voltage calculates the upper leg resistance. What do you think?

fabiobaltieri replied:
Oh, that would require specifying a `full-voltage` though. Maybe just return an error and keep the functionality for a followup.

fabiobaltieri commented:
> @galak @fabiobaltieri Hi, I've created a PR for macro and tests for the voltage-divider binding in this PR. Let me know what you guys think and if it looks good I'll stack the current-sensor on top of this PR.

Hey this is cool! Thanks for working on it, I think it's very useful, just a a comment on the macro definition and the actual usage of the binding.

ualbertagreen committed:
dts: bindings: adc: add shunt sensor

Add bindings for a current sensor using a shunt resistor.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen committed:
drivers: adc: current sense shunt

adds DT macro initializer and scaling function for current sense shunt.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen committed:
test: adc: current sense shunt

adds tests for the current sense shunt binding macro and rescaling
function.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen commented:
> > @galak @fabiobaltieri Hi, I've created a PR for macro and tests for the voltage-divider binding in this PR. Let me know what you guys think and if it looks good I'll stack the current-sensor on top of this PR.
> 
> Hey this is cool! Thanks for working on it, I think it's very useful, just a a comment on the macro definition and the actual usage of the binding.

Hi, I've edited the definitions and added the current sensor to this PR.

fabiobaltieri commented on include/zephyr/drivers/adc/adc_current_sense_shunt.h:
```diff
@@ -0,0 +1,56 @@
+ * @brief Calculates the actual voltage from the measured voltage
+ *
+ * @param[in] spec current sensor specification from Devicetree.
+ * @param[in,out] v_to_c Pointer to the measured voltage in millivolts on input, and the
+ * corresponding scaled current value in milliamps on output.
```
Should be v_to_i, conventional symbol for current is "I" (https://en.wikipedia.org/wiki/International_System_of_Units, https://en.wikipedia.org/wiki/SI_base_unit), C would be Coulomb.

fabiobaltieri commented on include/zephyr/drivers/adc/adc_current_sense_shunt.h:
```diff
@@ -0,0 +1,56 @@
+ * corresponding scaled current value in milliamps on output.
+ * @param[in] offset_mv offset voltage value used to shift the DC bias of the voltage sensor.
+ *
+ * @return 0 on success
```
Let's make this return void? Can't think of any special case where you may want to fail this.

fabiobaltieri commented on include/zephyr/drivers/adc/adc_current_sense_shunt.h:
```diff
@@ -0,0 +1,56 @@
+ * @param[in] spec current sensor specification from Devicetree.
+ * @param[in,out] v_to_c Pointer to the measured voltage in millivolts on input, and the
+ * corresponding scaled current value in milliamps on output.
+ * @param[in] offset_mv offset voltage value used to shift the DC bias of the voltage sensor.
```
Feels like `offset_mv` does not belong to here since it's not in the dt spec.

Should it be in the device tree spec and default to 0? Otherwise I'd suggest dropping it, it can always be applied to the caller.

ualbertagreen replied:
applying the offset to the caller sounds like a good idea. I've removed it from the function.

gmarull commented on dts/bindings/adc/current-sense-shunt.yaml:
```diff
@@ -0,0 +1,33 @@
+    description: |
+      Resistance of the shunt resistor in micro-ohms
+
+  gain:
+    type: int
+    default: 1
+    description: |
+      Gain used to amplify the voltage measured across the shunt resistor.
```
this is not clear to me, is this ADC gain? Also, defaults must be justified.

gmarull replied:
Ok, I see

ualbertagreen replied:
the analog gain is removed from the current-sense-shunt binding and put into the current-sense-amplifier binding instead following Linux convention.

gmarull commented on dts/bindings/adc/current-sense-shunt.yaml:
```diff
@@ -0,0 +1,33 @@
+  sense circuit.
+
+  This is based on Linux, documentation:
+    https://www.kernel.org/doc/Documentation/devicetree/bindings/iio/afe/current-sense-shunt.yaml
```
please place the binding in the same location to maintain compatibility

ualbertagreen replied:
Done.

gmarull reviewed (changes requested):
why isn't this implemented as a sensor? also see my comments

gmarull commented:
> > why isn't this implemented as a sensor? also see my comments
> 
> Hey, chiming in since I suggested this approach, so far the `voltage-divider` binding has been used just as a container and the fields picked at application level, figured that having an associated container struct, initializer and library style scaling function would make sense and would be a natural extension of `adc_raw_to_millivolts_dt`.
> 
> This could work as a potential driver but then strapping a dedicated device struct to this seems way overkill, and then you'd also have to handle the ADC sequence and read in the driver, which may simplify some case but then you are bound to do one channel at a time, while I would imagine most application to use ADC sequencer to do all channels if available and then do the scaling and further processing in the application. So in that sense I think it makes sense to have a good set of library function to make that easier while not binding to a specific setup.

I see, so it's more like a helper library on top of the base ADC library. If we keep things under drivers/adc, I think we should namespace things properly and also have docs in proper hierarchy, ie, ADC core > ADC Helpers {shunt, voltage-divider}.

gmarull commented on include/zephyr/drivers/adc/adc_current_sense_shunt.h:
```diff
@@ -0,0 +1,54 @@
+
+struct current_sense_shunt_dt_spec {
+	const struct adc_dt_spec port;
+	int shunt_micro_ohms;
+	int shunt_gain;
```
unsigned, probably fixed integers?

ualbertagreen replied:
Done.

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+
+struct voltage_divider_dt_spec {
+	const struct adc_dt_spec port;
+	int full_ohms;
+	int output_ohms;
```
ditto

ualbertagreen replied:
Done.

fabiobaltieri commented:
> I see, so it's more like a helper library on top of the base ADC interface. If we keep things under drivers/adc, I think we should namespace things properly and also have docs in proper hierarchy, ie, ADC core > ADC Helpers {shunt, voltage-divider}.

Yeah that's what I was thinking, the namespace could definite use some help, even having a bunch of files prefixed `adc_` when they are already in `include/zephyr/drivers/adc` seems unnecessary, maybe @anangl could chime in on that?

ualbertagreen committed:
dts: bindings: afe: add current sense amplifier

Add bindings for a current sensor using a shunt resistor and amplifier.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen committed:
drivers: afe: current sense amplifier

adds DT macro initializer and scaling function for current sense
amplifier.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

ualbertagreen committed:
test: adc: current sense amplifier

adds tests for the current sense amplifier binding macro and rescaling
function.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

fabiobaltieri reviewed (dismissed):
Neat, may be interesting to revisit the conversion functions down the road, I see Linux has some fancy logic using [`gcd()`](https://elixir.bootlin.com/linux/latest/source/drivers/iio/afe/iio-rescale.c#L335), Zephyr has a implementation of that in `lib/posix/getopt/getopt_long.c` whichi could be moved somewhere generic but I think this is plenty good enough for now.

sambhurst commented:
@gmarull Could you take another look when you have time. Thanks

MaureenHelm commented:
@gmarull @anangl please revisit

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+#define VOLTAGE_DIVIDER_DT_SPEC_GET(node_id)                                                       \
+	{                                                                                          \
+		.port = ADC_DT_SPEC_GET(node_id),                                                  \
+		.full_ohms = DT_PROP_OR(node_id, full_ohms, 0),                                    \
```
shouldn't `full-ohms` be required? full_ohms is >= output_ohms

fabiobaltieri replied:
From what I recall `full-ohms` was spec'd as optional so that the binding can be used to define a divider where the higher resistor is variable, like a PTC/NTC or something. I think we should leave it like this right now and in a followup PR handle that functionality, and maybe even extend it so that the fixed one can be the upper leg. Or we could require both and have a separate binding for that case, what do you think?

ualbertagreen replied:
The current `voltage-divider` documentation has `full-ohms` as optional to allow for `voltage-divider` to be used as a resistance based sensor instead:
https://docs.zephyrproject.org/latest/build/dts/api/bindings/adc/voltage-divider.html

I suppose we could change the binding to have a separate binding specifically for resistance based sensors, though I don't know if that's going to break any existing projects.

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+ * corresponding scaled voltage value on output.
+ *
+ * @return 0 on success
+ * @return ENOTSUP if "full_ohms" is not specified
```
`-E...`

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+
+#include <zephyr/drivers/adc.h>
+
+struct voltage_divider_dt_spec {
```
This isn't really a driver, so I wonder if this all belongs to `drivers/`...

fabiobaltieri replied:
should we start an `include/zephyr/adc` directory for non driver stuff?

gmarull commented on tests/drivers/adc/adc_rescale/src/main.c:
```diff
@@ -0,0 +1,100 @@
+
+ZTEST_USER(adc_rescale, test_get_adc_device)
+{
+	zassert_true(get_adc_device());
```
true? returns a struct device?

ualbertagreen replied:
Originally, I wanted to make sure `get_adc_device()` returns a real pointer and is not `NULL`, but looking at it again, the assert `zassert_true(device_is_ready(adc_dev), "ADC device is not ready");` in `get_adc_device()` already does that so the check is not necessary. I have removed it.

gmarull commented on tests/drivers/adc/adc_rescale/src/main.c:
```diff
@@ -0,0 +1,100 @@
+	zassert_true(get_adc_device());
+}
+
+static int init_adc(const struct adc_dt_spec *spec, int input_mv)
+{
+	int ret;
+
+	zassert_true(device_is_ready(spec->dev), "ADC device is not ready");
+
+	ret = adc_channel_setup_dt(spec);
+	zassert_equal(ret, 0, "Setting up of the first channel failed with code %d", ret);
```
why are we doing all this in a test? should the voltage divider API take care of ADC internally? I mean, given a "voltage-divider-dt-spec", I'd like to call something like `voltage_divider_get_voltage(&spec)`. How it's done internally, I don't care.

fabiobaltieri replied:
I think (sorry been a while from the first review) I suggested to leave the actual adc details outside of this to avoid dealing with the sequencer configuration: some application probably wants to have the sequencer do all measurement together and just convert, some do not support the sequencer at all.

gmarull reviewed (changes requested):
I still have some questions/concerns.

carlescufi commented:
@gmarull @fabiobaltieri please revisit

fabiobaltieri commented:
@gmarull can you do another pass? I think the comments have been addressed (or have a reply). Thanks!

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+ * corresponding scaled voltage value on output.
+ *
+ * @return 0 on success
+ * @return -ENOTSUP if "full_ohms" is not specified
```
`retval`

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+	return 0;
+}
+
+#endif
```
/* ZEPHYR_INCLUDE_DRIVERS_ADC_VOLTAGE_DIVIDER_H_ */ (to make things clear)

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+ * @return 0 on success
+ * @return -ENOTSUP if "full_ohms" is not specified
+ */
+static inline int divider_scale_voltage_dt(const struct voltage_divider_dt_spec *spec,
```
to be consistent, I'd use the same module namespacing everywhere, ie `voltage_divider_....`

gmarull commented on include/zephyr/drivers/adc/adc_voltage_divider.h:
```diff
@@ -0,0 +1,59 @@
+/*
```
nit: `s/adc_voltage_divider/voltage_divider/` (file name)

ualbertagreen replied:
Hi, for your other comments, I've made the other changes as suggested. 

I'm not sure what you mean with this comment, because later down you say to keep the drivers under `adc/`. Also I don't understand the advantage of putting each driver 2 directories deep just to hold the one file?

gmarull replied:
I mean to have `adc/voltage_divider.h` and not `adc/adc_voltage_divider.h` (ie double `adc`)

ualbertagreen replied:
The driver names have been modified to not duplicate `adc`.

gmarull commented on include/zephyr/drivers/adc/adc_current_sense_shunt.h:
```diff
@@ -0,0 +1,52 @@
+/*
```
ditto for file name

kartben commented on include/zephyr/drivers/adc/current_sense_amplifier.h:
```diff
@@ -0,0 +1,57 @@
+	}
+
+/**
+ * @brief Calculates the actual voltage from the measured voltage
```
s/voltage/amperage/

ualbertagreen replied:
I've opened a new PR to fix these errors #59892.

kartben commented on include/zephyr/drivers/adc/current_sense_shunt.h:
```diff
@@ -0,0 +1,52 @@
+	}
+
+/**
+ * @brief Calculates the actual voltage from the measured voltage
```
Calculates the actual *amperage*

kartben reviewed (commented):
@ualbertagreen I should probably open an issue, but are you able to please review some of the errors in the API doc comments?
